File: sdks/aperture-py/aperture_sdk/cache.py
import enum

from aperture_sdk._gen.aperture.flowcontrol.check.v1.check_pb2 import (
    HIT,
    CacheLookupStatus,
)


class LookupStatus(enum.Enum):
    HIT = "HIT"
    MISS = "MISS"


def convert_cache_lookup_status(status: CacheLookupStatus):
    return LookupStatus.HIT if status == HIT else LookupStatus.MISS


# convertCacheError converts a string error message to Python's Exception type.
# Returns None if the input string is empty.
def convert_cache_error(error_message):
    return None if not error_message else Exception(error_message)


# KeyLookupResponse is the interface that represents a cache value lookup.
class KeyLookupResponse:
    def __init__(self, value, lookup_status: LookupStatus, error):
        self.value = value
        self.lookup_status = lookup_status
        self.error = error

    def get_value(self):
        return self.value

    def get_lookup_status(self) -> LookupStatus:
        return self.lookup_status

    def get_error(self):
        return self.error


# KeyUpsertResponse represents response of updating or inserting a cache key.
class KeyUpsertResponse:
    def __init__(self, error):
        self.error = error

    def get_error(self):
        return self.error


# DeleteCachedValueResponse represents the response of deleting a cache key.
class KeyDeleteResponse:
    def __init__(self, error):
        self.error = error

    def get_error(self):
        return self.error

File: sdks/aperture-csharp/Core/Cache.cs
namespace ApertureSDK.Core;

using Aperture.Flowcontrol.Check.V1;

public enum LookupStatus
{
    HIT,
    MISS
}

public static class CacheLookupStatusConverter
{
    public static LookupStatus ConvertCacheLookupStatus(CacheLookupStatus status)
    {
        return status == CacheLookupStatus.Hit ? LookupStatus.HIT : LookupStatus.MISS;
    }
}

public static class CacheErrorConverter
{
    public static Exception? ConvertCacheError(string errorMessage)
    {
        return string.IsNullOrEmpty(errorMessage) ? null : new Exception(errorMessage);
    }
}

public class LookupResponse
{
    public object? Value { get; }
    public LookupStatus LookupStatus { get; }
    public Exception? Error { get; }

    public LookupResponse(object? value, LookupStatus lookupStatus, Exception? error)
    {
        Value = value;
        LookupStatus = lookupStatus;
        Error = error;
    }
}

public class UpsertResponse
{
    public Exception? Error { get; }

    public UpsertResponse(Exception? error)
    {
        Error = error;
    }
}

public class DeleteResponse
{
    public Exception? Error { get; }

    public DeleteResponse(Exception? error)
    {
        Error = error;
    }
}
